### PR TITLE
Feature: karmadactl init supports deployment through configuration files

### LINKED ISSUE #3464: Config file for karmadactl
**What would you like to be added**:

Config file for karmadactl

**Why is this needed**:

TODO

kubeadm config default:

TODO

```
apiVersion: kubeadm.k8s.io/v1beta3
bootstrapTokens:
- groups:
  - system:bootstrappers:kubeadm:default-node-token
  token: abcdef.0123456789abcdef
  ttl: 24h0m0s
  usages:
  - signing
  - authentication
kind: InitConfiguration
localAPIEndpoint:
  advertiseAddress: 1.2.3.4
  bindPort: 6443
nodeRegistration:
  criSocket: unix:///var/run/containerd/containerd.sock
  imagePullPolicy: IfNotPresent
  name: node
  taints: null
---
apiServer:
  timeoutForControlPlane: 4m0s
apiVersion: kubeadm.k8s.io/v1beta3
certificatesDir: /etc/kubernetes/pki
clusterName: kubernetes
controllerManager: {}
dns: {}
etcd:
  local:
    dataDir: /var/lib/etcd
imageRepository: registry.k8s.io
kind: ClusterConfiguration
kubernetesVersion: 1.26.0
networking:
  dnsDomain: cluster.local
  serviceSubnet: 10.96.0.0/12
scheduler: {}
```

### PR DESCRIPTION
**Pull request by tiansuo114**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Increased karmadactl's ability to init by reading configuration files
From
```shell
user@virtual-machine:karmada/_output/bin/linux/amd64# karmadactl init --karmada-apiserver-image registry.k8s.io/kube-apiserver:v1.30.0 --karmada-kube-controller-manager-image registry.k8s.io/kube-controller-manager:v1.30.0 --etcd-image registry.k8s.io/etcd:3.5.13-0 --etcd-init-image docker.io/library/alpine:3.19.1 --karmada-aggregated-apiserver-image docker.io/karmada/karmada-aggregated-apiserver:v1.10.3 --karmada-controller-manager-image docker.io/karmada/karmada-controller-manager:v1.10.3 --karmada-scheduler-image docker.io/karmada/karmada-scheduler:v1.10.3 --karmada-webhook-image docker.io/karmada/karmada-webhook:v1.10.3 --crds /home/tiansuo/Downloads/crds.tar.gz ....
```
Turn into
```shell
user@virtual-machine:karmada/_output/bin/linux/amd64# karmadactl init --config config.yaml
```
Example of config.yaml
```yaml
apiVersion: config.karmada.io/v1alpha1
kind: KarmadaInitConfig
metadata:
  name: karmada-init
spec:
  karmadaCrds: "https://github.com/karmada-io/karmada/releases/download/v1.10.3/crds.tar.gz"
  etcd:
    local:
      imageRepository: "registry.k8s.io/etcd"
      imageTag: "3.5.13-0"
      initImage:
        imageRepository: "docker.io/library/alpine"
        imageTag: "3.19.1"
  components:
    karmadaAPIServer:
      imageRepository: "registry.k8s.io/kube-apiserver"
      imageTag: "v1.30.0"
    karmadaAggregatedAPIServer:
      imageRepository: "docker.io/karmada/karmada-aggregated-apiserver"
      imageTag: "v1.10.3"
    kubeControllerManager:
      imageRepository: "registry.k8s.io/kube-controller-manager"
      imageTag: "v1.30.0"
    karmadaControllerManager:
      imageRepository: "docker.io/karmada/karmada-controller-manager"
      imageTag: "v1.10.3"
    karmadaScheduler:
      imageRepository: "docker.io/karmada/karmada-scheduler"
      imageTag: "v1.10.3"
    karmadaWebhook:
      imageRepository: "docker.io/karmada/karmada-webhook"
      imageTag: "v1.10.3"

```

**Which issue(s) this PR fixes**:
Fixes #3464 
This PR is related to the documentation PR: https://github.com/karmada-io/karmada/pull/5277
**Special notes for your reviewer**:
@liangyuanpeng 
**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by liangyuanpeng** [comment]
Hi, this is a bot comment, just put the label of `ok-to-test` here, you can comment `/retest` to rerun github workflow if github workflow is failing and it's not releated with this PR. 

/ok-to-test

File: pkg/karmadactl/cmdinit/config/config_test.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/config/config_test.go:27]:
```suggestion
apiVersion: config.karmada.io/v1alpha1
```

I think the first version is `v1alpha1`

**Comment by liangyuanpeng** [comment]
please link to the proposal at PR describe. 

`releated #5277`

**Comment by tiansuo114** [comment]
This PR is related to the documentation PR: [#5277 ](https://[github.com/user/repo/pull/123](https://github.com/karmada-io/karmada/pull/5277))

**Comment by tiansuo114** [comment]
> please link to the proposal at PR describe.请链接到 PR 描述中的提案。
> 
> `releated #5277`

Ok, I've now connected all three related PRS to document pr

**Comment by tiansuo114** [comment]
Can we start the review of this pr first, or do we still need to wait for the merge of [docs pr](https://github.com/karmada-io/karmada/pull/5277) before we can conduct the review of this pr

**Comment by tiansuo114** [comment]
Can we start the review of this pr first, or do we still need to wait for the merge of https://github.com/karmada-io/karmada/pull/5277 before we can conduct the review of this pr
@liangyuanpeng

**Comment by liangyuanpeng** [comment]
I think we need to merge proposal first

**Comment by tiansuo114** [comment]
The code in this branch has been modified to use the configuration file format from the existing proposal for deployment, and it has been verified that the deployment works correctly. Additionally, the proposal has already been merged into the master branch. Can we start the review now?
@liangyuanpeng 

Furthermore, I have some questions in another [PR](https://github.com/karmada-io/karmada/pull/5623#issuecomment-2399002327), and we might need to discuss them further.

File: pkg/karmadactl/cmdinit/config/config.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/config/config.go:57]: is it possible to use viper to load config file, like

```
workDir, _ := os.Getwd()
	viper.SetConfigName("config")
	viper.SetConfigType("yml")
	viper.AddConfigPath(path.Join(workDir, "./"))
	//viper.AddConfigPath(path.Join(workDir, "config/application.yml"))
	err := viper.ReadInConfig()
```

File: pkg/karmadactl/cmdinit/config/config_test.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/config/config_test.go:47]: `karmada/etcd` does not seem to be an efficient container repository

File: pkg/karmadactl/cmdinit/config/config.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config.go:57]: Using viper for parsing is indeed a better choice, and I will change it soon

File: pkg/karmadactl/cmdinit/config/config_test.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config_test.go:47]: The repository field here is an error, my test case documentation here is automatically generated, there are some errors, I will correct

File: pkg/karmadactl/cmdinit/config/config.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config.go:57]: Here, using Viper might not be the best choice. I am still researching how to better use the Kubernetes API library to handle different versions of configuration files. It seems that Viper may not be particularly suitable for this scenario. I would appreciate more of your advice.

File: pkg/karmadactl/cmdinit/config/config_test.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config_test.go]: The problem here has been modified, and I hope to get your further guidance

File: pkg/karmadactl/cmdinit/config/config.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config.go:57]: I think it's a better choice to use k8s.io/apimachinery/pkg/runtime/serializer/yaml for parsing here. What do you think? I have tried using Viper for parsing, but the results were not very satisfactory. Using the k8s.io/apimachinery/pkg/runtime/serializer/yaml package for parsing might also facilitate future modifications to support parsing methods for multiple versions of configuration files. What do you think about keeping it as is without making changes here?

Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config.go:57]: I think it's a better choice to use `k8s.io/apimachinery/pkg/runtime/serializer/yaml` for parsing here. What do you think? I have tried using Viper for parsing, but the results were not very satisfactory. Using the `k8s.io/apimachinery/pkg/runtime/serializer/yaml` package for parsing might also facilitate future modifications to support parsing methods for multiple versions of configuration files. What do you think about keeping it as is without making changes here?
@liangyuanpeng

File: pkg/karmadactl/cmdinit/config/config_test.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config_test.go]: The problem here has been modified, and I hope to get your further guidance
@liangyuanpeng

File: .github/workflows/cli.yaml
Comment (liangyuanpeng) [inline .github/workflows/cli.yaml:45]: Adding this CI is in the right direction, but I guess we need to add it in another pr to reduce the size of each pr,And let the CI stand alone in a separate action, rather than with the run karmadactl init test

File: hack/cli-testing-init-with-config.sh
Comment (liangyuanpeng) [inline hack/cli-testing-init-with-config.sh:38]:
```suggestion
kind_version=v0.24.0
```

File: hack/cli-testing-init-with-config.sh
Comment (liangyuanpeng) [inline hack/cli-testing-init-with-config.sh:134]:
```suggestion
    kubeconfig: "${KUBECONFIG_PATH}/${HOST_CLUSTER_NAME}.config"  
  components:
    karmadaControllerManager:
      imageRepository: "${REGISTRY}/karmada-controller-manager"    
      imageTag: "${VERSION}"                                       
    karmadaScheduler:
      imageRepository: "${REGISTRY}/karmada-scheduler"             
      imageTag: "${VERSION}"                                       
    karmadaWebhook:
      imageRepository: "${REGISTRY}/karmada-webhook"               
      imageTag: "${VERSION}"                                      
    karmadaAggregatedAPIServer:
      imageRepository: "${REGISTRY}/karmada-aggregated-apiserver"  
      imageTag: "${VERSION}"                                       
  karmadaDataPath: "${HOME}/karmada"                               
  karmadaPkiPath: "${HOME}/karmada/pki"                            
  karmadaCrds: "./crds.tar.gz"                                     
```

File: pkg/karmadactl/cmdinit/config/config.go
Comment (liangyuanpeng) [reply pkg/karmadactl/cmdinit/config/config.go:57]: It will be helpful to link the kubeadm parsing configuration file code here

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:386]:
```suggestion
```

**Review by liangyuanpeng — COMMENTED** [review]
sorry for my late and add some comment

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/kubernetes/deploy.go:905]: Have you seen how kubeadm does this?

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:765]: Yes, there is an error here, and I will fix it immediately.

File: .github/workflows/cli.yaml
Comment (tiansuo114) [reply .github/workflows/cli.yaml:45]: Okay, I will isolate it as a separate PR and place it in a distinct CI section.

File: pkg/karmadactl/cmdinit/config/config.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config.go:57]: In kubeadm, the parsing of configuration files is done by maintaining a global CodecFactory, which registers different version-specific configuration file decoders within its runtime.Decoder. This approach allows for the parsing of configuration files corresponding to different versions, as shown in the code linked here
https://github.com/kubernetes/kubernetes/blob/4f796c02f77fb95d42cd161ea663dd1bf05e372f/cmd/kubeadm/app/util/config/initconfiguration.go#L338C1-L347C4
https://github.com/kubernetes/kubernetes/blob/4f796c02f77fb95d42cd161ea663dd1bf05e372f/cmd/kubeadm/app/apis/kubeadm/v1beta3/register.go#L31C1-L64C2

As for how the configuration file is translated into the information used in the structures for actual deployment operations, I will address that in a comment below regarding that specific information.

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go:905]: In kubeadm, this step is automatically completed by some code generation tools provided by the Kubernetes official. The tool used here is conversion-gen, and the specific code link is as follows (https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta3/zz_generated.conversion.go). The reason I did not use the code generated by this tool is that the automatically generated conversion methods are related to the structure of the structs, JSON fields, and the field names of the elements within the structs. Since the CommandInitOption structure in karmadactl differs significantly from the KarmadaInitConfig structure used for parsing the configuration file, it seems unsuitable to use the code generation tool to generate the code.

File: pkg/karmadactl/cmdinit/kubernetes/deploy.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go]: The issues mentioned above have been resolved, and I look forward to your further guidance 🌼

Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/kubernetes/deploy.go]: The issues mentioned above have been resolved, and I look forward to your further guidance 🌼
@liangyuanpeng

**Comment by RainbowMango** [comment]
> This PR first reads the configuration file and then convert to CommandInitOption, which is compatible with the legacy code unless you refactor the legacy code logic. Do you agree?

Yeah, I agree.

**Comment by zhzhuang-zju** [comment]
> > This PR first reads the configuration file and then convert to CommandInitOption, which is compatible with the legacy code unless you refactor the legacy code logic. Do you agree?

agree

**Comment by tiansuo114** [comment]
> This PR first reads the configuration file and then convert to `CommandInitOption`, which is compatible with the legacy code unless you refactor the legacy code logic. Do you agree?
> 
> cc @zhzhuang-zju @RainbowMango

Yes, the code here should be fully compatible with the previous code because the `CommandInitOption` structure is still being used in the actual deployment, while the `KarmadaInitConfig` structure is only used for reading the configuration file.

File: pkg/karmadactl/cmdinit/config/config.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config.go:52]: Yes, the methods here can be integrated into a single method. When I initially designed it, I aimed to refer to the function execution logic of kubeadm as much as possible, which is why it was designed in this structure.

File: pkg/karmadactl/cmdinit/config/types.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/types.go]: The modification of these parts has been completed, and we look forward to your further guidance
@liangyuanpeng

Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/types.go]: The modification of these parts has been completed, and the code has passed github ci, looking forward to your further guidance
@liangyuanpeng

File: go.mod
Comment (RainbowMango) [inline go.mod:21]: The `github.com/pkg/errors` has been archived, we should not use it anymore.

**Comment by RainbowMango** [comment]
@tiansuo114 Please help to squash commits, and here is the guide line: https://karmada.io/docs/contributor/github-workflow/#squash-commits.
I can help to take a look tomorrow.

**Comment by tiansuo114** [comment]
> @tiansuo114 Please help to squash commits, and here is the guide line: https://karmada.io/docs/contributor/github-workflow/#squash-commits.请帮助压缩提交，这里是指导方针： [https](https://karmada.io/docs/contributor/github-workflow/#squash-commits) ://karmada.io/docs/contributor/github-workflow/#squash-commits。 I can help to take a look tomorrow.明天我可以帮忙看看。

At first, I wasn't sure whether to squash commits as I go or to keep each stage's commits and squash them all after the review to make it easier to track changes at different stages. I will start by doing a squash and wait for everyone's review.

**Review by RainbowMango — COMMENTED** [review]
/assign

File: pkg/karmadactl/cmdinit/config/types.go
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/config/types.go:58]:
```suggestion
	// KarmadaCRDs configures the Karmada CRDs to be installed
	// +optional
	KarmadaCRDs string `json:"karmadaCRDs,omitempty" yaml:"karmadaCRDs,omitempty"`
```

**Review by RainbowMango — COMMENTED** [review]
Generally looks good to me after a quick go-through.  Just some nits about naming convention.
@liangyuanpeng do you have any further comments?
@tiansuo114 Have you tested it? Can you share the test report here?

File: pkg/karmadactl/cmdinit/config/types.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/types.go]: Issues with naming conventions here have been fixed

**Comment by tiansuo114** [comment]
> Generally looks good to me after a quick go-through. Just some nits about naming convention. @liangyuanpeng do you have any further comments? @tiansuo114 Have you tested it? Can you share the test report here?

I conducted some tests using kind cluster in a virtual machine on my local machine. I also attempted to add a test task in the CLI section of Karmada's GitHub CI workflow to verify whether the karmadactl init --init-config command runs correctly. (The detailed testing process can first be observed in this [link](https://github.com/tiansuo114/karmada/actions/runs/11456237037))Following the guidance of my mentor, I plan to submit a separate PR to merge the test process into the main branch after the current PR is merged.

**Comment by RainbowMango** [comment]
OK.
/approve
Leave LGTM to @liangyuanpeng for final decision.

File: pkg/karmadactl/cmdinit/cmdinit.go
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/cmdinit.go:82]:
```suggestion
		%[1]s init --init-config /path/to/your/config/file.yaml`)
```
Let's use `tab` to keep the original format aligned.

File: pkg/karmadactl/cmdinit/cmdinit.go
Comment (RainbowMango) [inline pkg/karmadactl/cmdinit/cmdinit.go:155]:
```suggestion
	flags.StringVar(&opts.KarmadaInitFilePath, "config", "", "Karmada init file path")
```
@liangyuanpeng I think `--config` should be fine, because its under `init`, so we can use a shorter name.
`karmadatl init --config`.

File: pkg/karmadactl/cmdinit/cmdinit.go
Comment (RainbowMango) [reply pkg/karmadactl/cmdinit/cmdinit.go]: We can skip the `init` prefix as it's inside the `init` command. See my comments at https://github.com/karmada-io/karmada/pull/5357#discussion_r1811726710

File: pkg/karmadactl/cmdinit/cmdinit.go
Comment (liangyuanpeng) [reply pkg/karmadactl/cmdinit/cmdinit.go:155]: oops, sorry about that, I confused it with `karmadactl config image xxx`, please help restore `init-config` to `config`
@tiansuo114

File: pkg/karmadactl/cmdinit/config/config.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/config/config.go:60]: I guss we  need some test for it

File: pkg/karmadactl/cmdinit/config/config_test.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/config/config_test.go:130]: What happens when I pass `config.karmada.io/v1`, maybe add some test for it.

Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config_test.go:132]: In this section, the YAML file's group, version, and kind will be checked to see if they match the current configuration file format. If they don't match, the deployment will not proceed, and an error will be returned. Unit tests for this part have already been added to pkg/karmadactl/cmdinit/config/config_test.go.

File: pkg/karmadactl/cmdinit/config/config.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/config/config.go:60]: Alright, the tests for ParseGVKYamlMap have been added to `pkg/karmadactl/cmdinit/config/config_test.go`.

File: pkg/karmadactl/cmdinit/cmdinit.go
Comment (tiansuo114) [reply pkg/karmadactl/cmdinit/cmdinit.go]: OK,the changes in these areas have been completed, and now we are waiting for the GitHub CI process to run.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

Hi @liangyuanpeng Let's move this forward and handle further comments by separated PRs.

**Comment by liangyuanpeng** [comment]
Hi, this is a bot comment, just put the label of `ok-to-test` here, you can comment `/retest` to rerun github workflow if github workflow is failing and it's not releated with this PR. 

/ok-to-test

File: pkg/karmadactl/cmdinit/config/types.go
Comment (liangyuanpeng) [inline pkg/karmadactl/cmdinit/config/types.go:34]: I think this is not need, let's delete it.

Comment (RainbowMango) [reply pkg/karmadactl/cmdinit/config/types.go:34]: Do you mean the `metav1.ObjectMeta`? It makes sense to me.
Please add a task to the tracking issue.

Comment (liangyuanpeng) [reply pkg/karmadactl/cmdinit/config/types.go:34]: Updated,  How about including it in https://github.com/karmada-io/karmada/pull/5306#issuecomment-2441115638 (for Summer OSPP 2024)?,  or alone PR for it

Comment (RainbowMango) [reply pkg/karmadactl/cmdinit/config/types.go:34]: Both are OK to me. I just want somewhere track this.

File: hack/verify-license.sh
Comment (RainbowMango) [inline hack/verify-license.sh:43]: @tiansuo114 Can you help to recall why need to ignore `.git` here?

PS: I'm exploring the reason why the license check becomes inactive, and seems this line breaks the functionality.

Comment (tiansuo114) [reply hack/verify-license.sh:43]: In my detailed response under https://github.com/karmada-io/karmada/pull/5984#issuecomment-2562447851, the idea is that files under the `.git/` directory, similar to those under `.github/`, `.idea/`, and other paths, should not be subject to license checks. When I ran the related check scripts locally, I encountered issues related to this, so I added path ignore rules to the script.

**Pull request merged**
